### ISSUE TITLE
cpu_only mode unnecssary copy

### ISSUE THREAD

**Issue by tnowotny** [genn-team/genn#56]
At the moment one needs to make a differently named copy from the .cu file(s) so that the compiler is not confused by the suffix.
So, in cpu_only mode, one, for example, needs to create something like classol_sim_cpu_only.cc from classol_sim.cu to get this done. Note, it has to be different in suffix and base name. This should be improved in the future so that "make cpu_only" suffices to sort this out.

**Comment by Sayantan2048** [comment]
Hello sir, the cpu_only mode currntly depends on CUDA API. For that reason, users without cuda toolkit cannot run genn using cpu_only. As part of my GSoC proposl, should we make a CPU only mode that runs natively on CPU without CUDA/OpenCL? Also native CPU mode opens more possiblites for optimizations on hardware like Intel MIC.

**Comment by tnowotny** [comment]
Hi Sayantan,
CPU only mode has always been intended to be CUDA independent. In my tests that was successful but I may have missed something. Where is the remaining CUDA dependence when CPU_only mode is compiled?

**Comment by Sayantan2048** [comment]
Hello Sir,
I removed CUDA_PATH and then I ran the following userproject from project directory:
`make clean && make
./generate_run 0 100 1000 20 100 0.0025 outname MBody_delayedSyn 0 FLOAT 0
`
I got the following error message i.e the first error reported by compiler:
`/home/sayantan/genn/lib/include/global.h:32:26: fatal error: cuda_runtime.h: No such file or directory`

I went through the control flow and found that the code generation is CUDA dependent for CPU and GPU. For examle chooseDevice() function in generateALL cannot be compiled without CUDA host APIs. 

However, on a remote machine with properly installed CUDA and CUDA_PATH unharmed, I was able to run the above model on CPU. I looked into the generated neuronFnct.cc and runner.cc and found that runner.cc had some dependence on CUDA host API functions.

So, to sum up, it looks like both code generation and generated code are CUDA host API dependent even for CPU.

**Comment by tnowotny** [comment]
Hi - I am sorry for the confusion but what you get when using the generate_run examples and choosing execution on the CPU is never the CPU_ONLY mode that this issue refers to. I know it's a bit confusing but the philosophy of GeNN has always been that both CPU and GPU code are available at compile time and the user is free to program their user side code to use either or both (which the program invoked by generate_run does).
However, there is also a proper CPU_ONLY mode in the development branch for the case where CUDA or appropriate hardware are not available. Checkout the development branch and have a look at the CPU_ONLY option, e.g. in the MBody1 user project's generate_run.cc.
Using this option, all dependencies on CUDA should be removed; however the renaming problem for classol_sim.cu which started this issue, still remains because even if CUDA is not installed the .cu ending is interpreted by the c++ compiler.

**Comment by jamesturner246** [comment]
Changes merged in #65. I'll post the same thing I posted there:

You should only need one GeNN directory for all work really... A user's own project can be in any directory, say $HOME/my_project, and they all include the same includes from the same GeNN tree. The only use-case I can think of for having multiple GeNN trees is if you're an active developer.

In my post in your new thread ( #67 ), I suggest the use of environment 'modules', where switching  GENN_PATH and PATH etc. can be achieved using, for example: 
module switch [genn_module_1] [genn_module_2]

What do you all think?
James.

**Comment by tnowotny** [comment]
One thought: When working in an HPC environment, where several jobs will run in parallel and without control when the code generation stage happens, there can be collisions of compiles in the GeNN directory. In that case, working in separate copies has been useful. However, I think, ideally, there should be one GeNN copy and we should refactor the system so that all builds go into the current user project directory (instead of building e.g. generate_all "in place" in the GeNN source directory), so that all conflicts of this kind can be avoided while the GeNN install can stay unique and available for any running project.

**Comment by jamesturner246** [comment]
We could use a little script to copy the userproject dir to the user's
home, like CUDA does with its samples. Then generateALL can be compiled
into the sample project directory, or user's project directory, as you
say. Then all the cluster admin has to worry about is loading a GeNN
module to load all the environment variables correctly.

I'll see what I can do.
James.

**Comment by Sayantan2048** [comment]
Alternative to switching modules, we can also use:

`source <(./set_path.sh)`

where `cat set_path.sh` is

```
GENN_PATH=$(readlink -f $(dirname $0))
echo "export GENN_PATH=$GENN_PATH"
echo "export PATH=$PATH:$GENN_PATH/lib/bin"
```

**Comment by jamesturner246** [comment]
These issues are since solved in the development branch, pre-GeNN 2.2 release.
